File: src/main/java/jukebot/commands/Help.java
package jukebot.commands;

import jukebot.Database;
import jukebot.EventListener;
import jukebot.JukeBot;
import jukebot.utils.Command;
import jukebot.utils.CommandProperties;
import jukebot.utils.Helpers;
import net.dv8tion.jda.core.EmbedBuilder;
import net.dv8tion.jda.core.entities.MessageEmbed;
import net.dv8tion.jda.core.events.message.guild.GuildMessageReceivedEvent;

import java.util.Comparator;
import java.util.function.Predicate;
import java.util.stream.Stream;

@CommandProperties(description = "Displays all commands", aliases = {"commands"}, category = CommandProperties.category.MISC)
public class Help implements Command {

    public void execute(GuildMessageReceivedEvent e, String query) {

        switch (query) {
            case "1":
                e.getChannel().sendMessage(CreateHelpEmbed(
                        "You can use the **play** command to make JukeBot join your channel, search for the specified song and begin playing.\n`"
                                + Database.getPrefix(e.getGuild().getIdLong()) + "play <URL/Search Query>`"
                )).queue();
                break;

            case "2":
                final StringBuilder controls = new StringBuilder();

                filterCommands(command -> command.properties().category() == CommandProperties.category.CONTROLS).forEach(command ->
                        controls.append("**`").append(Helpers.padRight(" ", command.name().toLowerCase(), 11)).append(":`** ")
                                .append(command.properties().description()).append("\n")
                );

                e.getChannel().sendMessage(CreateHelpEmbed(controls.toString())).queue();
                break;

            case "3":
                final StringBuilder media = new StringBuilder();

                filterCommands(command -> command.properties().category() == CommandProperties.category.MEDIA).forEach(command ->
                        media.append("**`").append(Helpers.padRight(" ", command.name().toLowerCase(), 11)).append(":`** ")
                                .append(command.properties().description()).append("\n")
                );

                e.getChannel().sendMessage(CreateHelpEmbed(media.toString())).queue();
                break;

            case "4":
                final StringBuilder misc = new StringBuilder();

                filterCommands(command -> command.properties().category() == CommandProperties.category.MISC).forEach(command ->
                        misc.append("**`").append(Helpers.padRight(" ", command.name().toLowerCase(), 11)).append(":`** ")
                                .append(command.properties().description()).append("\n")
                );

                e.getChannel().sendMessage(CreateHelpEmbed(misc.toString())).queue();
                break;

            default:
                e.getChannel().sendMessage(new EmbedBuilder()
                        .setColor(JukeBot.embedColour)
                        .setTitle("Help Categories")
                        .setDescription("`1.` Getting Started\n`2.` Controls\n`3.` Media\n`4.` Miscellaneous\n\nUse `" + Database.getPrefix(e.getGuild().getIdLong()) + "help <number>` to select a category")
                        .build()
                ).queue();
        }
    }

    private Stream<Command> filterCommands(Predicate<Command> filter) {
        return EventListener.commands.values()
                .stream()
                .filter(filter)
                .sorted(Comparator.comparing(Command::name));
    }

    private MessageEmbed CreateHelpEmbed(String description) {
        return new EmbedBuilder()
                .setColor(JukeBot.embedColour)
                .setDescription("[View more information here](http://jukebot.xyz/documentation)\n" + description)
                .build();
    }

}

File: src/main/java/jukebot/commands/Help.kt
package jukebot.commands

import jukebot.listeners.CommandHandler
import jukebot.framework.Command
import jukebot.framework.CommandCategory
import jukebot.framework.CommandProperties
import jukebot.framework.Context

@CommandProperties(description = "Displays all commands", aliases = ["commands", "cmds", "?"])
class Help : Command(ExecutionType.STANDARD) {

    private val categories = CommandCategory.values()
            .mapIndexed { i, e -> "`${i + 1}.` **`${pad(e.toTitleCase())}:`** ${e.description}" }
            .joinToString("\n")

    fun pad(s: String): String {
        return String.format("%-12s", s).replace(" ", " \u200B")
    }

    override fun execute(context: Context) {
        if (context.args.isEmpty()) {
            return sendDefaultHelp(context)
        }

        val menu = context.args.firstOrNull()?.toIntOrNull() ?: 0

        if (menu <= 0 || menu > CommandCategory.values().size) {
            val cmd = CommandHandler.commands
                    .filter { it.key == context.args[0] || it.value.properties().aliases.contains(context.args[0]) }
                    .values
                    .firstOrNull()
                    ?: return sendDefaultHelp(context)

            sendCommandHelp(context, cmd)
        } else {
            val category = CommandCategory.values()[menu - 1]
            val builder = StringBuilder()

            for (cmd in commandsByCategory(category)) {
                builder.append("**`")
                        .append(pad(cmd.name().toLowerCase()))
                        .append(":`** ")
                        .append(cmd.properties().description)
                        .append("\n")
            }

            context.embed {
                setTitle("Help for **$category**")
                setDescription("[View more information here](https://jukebot.serux.pro/documentation)\n$builder")
                setFooter("You can use ${context.prefix}help <command> to view additional command information", null)
            }
        }
    }

    fun sendDefaultHelp(ctx: Context) {
        ctx.embed {
            setColor(ctx.embedColor)
            setTitle("JukeBot Help Menu")
            setDescription("Get started by joining a voicechannel and sending `${ctx.prefix}play <query>`!")
            addField("Categories", categories, true)
            addField("Links", "[Discord](https://discord.gg/xvtH2Yn) | [Website](https://jukebot.serux.pro)", true)
            setFooter("Select a category with ${ctx.prefix}help <number>", null)
        }
    }

    fun sendCommandHelp(context: Context, cmd: Command) {
        val aliases = cmd.properties().aliases
        val aliasString = if (aliases.isEmpty()) "None" else aliases.joinToString(", ")

        context.embed("Help for **${cmd.name()}**",
                "**Aliases:** $aliasString\n**Description:** ${cmd.properties().description}")
    }

    private fun commandsByCategory(category: CommandCategory) =
            CommandHandler.commands
                    .values
                    .filter { it.properties().category == category }
                    .sortedBy { it.name() }

}
